dvcfs: initial docs (#3932)
iterative/dvc.org: content/docs/api-reference/dvcfilesystem.md

# DvcFileSystem

> _New in DVC 2.26_

Author: ```suggestion
_New in DVC 2.27_
```

> DVCFileSystem provides a pythonic file interface (
> [fsspec-compatible](https://filesystem-spec.readthedocs.io/)) for a DVC repo. It
> is a read-only filesystem, hence it does not support any write operations, like
> `put_file`, `cp`, `rm`, `mv`, `mkdir` etc.

Reviewer: 💅🏼 
```suggestion
`DVCFileSystem` is a Python class that provides a file interface for <abbr>DVC repositories</abbr>
(compatible with [fsspec](https://filesystem-spec.readthedocs.io/)). It
is read-only, hence it does not support any write operations like
`put_file`, `cp`, `rm`, `mv`, `mkdir` etc.
```

> DVCFileSystem provides a unified view of all the files/directories in your
> repository, be it Git-tracked or DVC-tracked, or untracked (in case of a local
> repository). It can reuse the files in DVC <abbr>cache</abbr> and can otherwise
> stream from
> [supported remote storage](/doc/command-reference/remote/add#supported-storage-types).

Reviewer: ```suggestion
`DVCFileSystem` objects construct a unified view of all the files and directories in your
repo, be it Git-tracked, DVC-tracked, or untracked (in the case of local
repos). It can reuse files in the <abbr>DVC cache</abbr>, and it can otherwise
stream from [supported remote storage].

[supported remote storage]: /doc/command-reference/remote/add#supported-storage-types
```

> It can reuse files in the DVC cache

What does this mean though? Reuse them for what purpose?

> ```py
> >>> from dvc.api import DVCFileSystem
> # opening a local repository
> >>> fs = DVCFileSystem("/path/to/local/repository")
> # opening a remote repository
> >>> url = "https://github.com/iterative/example-get-started.git"
> >>> fs = DVCFileSystem(url, rev="main")

Reviewer: 💅🏼 
````suggestion
```py
from dvc.api import DVCFileSystem
# Opening a local repository
fs = DVCFileSystem("/path/to/local/repository")
# Opening a remote repository
url = "https://github.com/iterative/example-get-started.git"
fs = DVCFileSystem(url, rev="main")
````

```

The optional positional argument can be a URL or a local path to the DVC

> project. If unspecified, the DVC project in current working directory is used.

Reviewer: Does it work with any DVC project or does it have to be a Git repo? (We currently use the term "repo" specifically in a few places.)

The optional `rev` argument can be passed to open a filesystem from a certain
Git commit (any [revision](https://git-scm.com/docs/revisions) such as a branch
or a tag name, a commit hash, or an [experiment name]).

> [experiment name]: /doc/command-reference/exp/run#-n

Reviewer: Would it be better to put the [class signature](https://github.com/iterative/dvc/blob/391bb29be29fe9ac3036180f503590c97b71c850/dvc/fs/dvc.py#L368) up somewhere? With param names, etc. (may need to make a mixed definition combined with [FileSystem](https://github.com/iterative/dvc/blob/570e3a5359b41520ea01952df11602e40d69f793/dvc/objects/fs/base.py#L55))

    '/evaluation/plots/confusion_matrix.json',
    '/evaluation/plots/precision_recall.json',
    '/evaluation/plots/roc.json',
    '/model.pkl',
    ...
]
```

This is similar to `dvc ls --recursive` CLI command. It returns all of the files
tracked by DVC and Git and if filesystem is opened locally, it also includes the
local untracked files.

## Downloading a file or a directory

```py
>>> fs.get_file("data/data.xml", "data.xml")
```

This downloads "data/data.xml" file to the current working directory as
"data.xml" file. The DVC-tracked files may be downloaded from the cache if it
exists or may get streamed from the remote.

```py
>>> fs.get("data", "data", recursive=True)
```

This downloads all the files in "data" directory - be it Git-tracked or
DVC-tracked into a local directory "data". Similarly, DVC might fetch files from
remote if they don't exist in the cache.

## API Reference

As DVCFileSystem is based on [fsspec](https://filesystem-spec.readthedocs.io/),
it is compatible with most of the APIs that it offers. Please check the fsspec's
[API Reference](https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.spec.AbstractFileSystem)

> for more details.

Reviewer: Last note: we could add the term "methods" or something here, to clarify we mean the reference of the elements of the class/object, not some larger API.
